mockatests/drivers/test-scheduler.c

```diff
@@ -0,0 +1,1315 @@
+/*
+ * Copyright (c) 2024, Cloud Software Group, Inc.
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived from
+ *     this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER
+ * OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#define _GNU_SOURCE /* Needed for F_GETPIPE_SZ */
+
+#include <setjmp.h>
+#include <stdarg.h>
+#include <stddef.h>
+#include <cmocka.h>
+
+#include <errno.h>
+#include <limits.h>
+#include <stdio.h>
+#include <sys/time.h>
+
+#include "scheduler.c"
+
+struct timeval fake_gettimeofday;
+int __wrap_gettimeofday(struct timeval* tv, struct timezone* tz)
+{
+  *tv = fake_gettimeofday;
+  return 0;
+}
+
+void scheduler_gc_events(scheduler_t *s);
+
+typedef struct {
+  bool was_called;
+  char mode;
+  event_id_t id;
+} event_cb_spy_t;
+
+void mock_event_cb(event_id_t id, char mode, void *private)
+{
+  event_cb_spy_t* out = (event_cb_spy_t*)private;
+  out->was_called = true;
+  out->mode = mode;
+  out->id = id;
+}
+
+static int
+event_queue_length(const scheduler_t* s)
+{
+  int len = 0;
+  event_t *event;
+  scheduler_for_each_event(s, event){
+    len += 1;
+  }
+  return len;
+}
+
+void fake_event_cb (event_id_t id, char mode, void *private) {}
+
+void
+test_scheduler_set_max_timeout(void **state)
```

Comment:
`test_scheduler_set_max_timeout_negative` ?

Reply:
Added in b5a2f60.